Data/HashMap/Base.hs

```diff
@@ -1883,6 +1890,33 @@ fromListWith :: (Eq k, Hashable k) => (v -> v -> v) -> [(k, v)] -> HashMap k v
 fromListWith f = L.foldl' (\ m (k, v) -> unsafeInsertWith f k v m) empty
 {-# INLINE fromListWith #-}
 
+-- | /O(n*log n)/ Construct a map from a list of elements.  Uses
+-- the provided function to merge duplicate entries.
+--
+-- === Examples
+--
+-- Given a list of key-value pairs where the keys are of different flavours, e.g:
+--
+-- > data Key = Mul | Add
+--
+-- and the values need to be combined differently on a collision, depending on the key:
+--
+-- > combine Mul = (*)
+-- > combine Add = (+)
```

Comment:
Commutative operations make it harder to infer the behaviour of the function from the example.

Reply:
Really? I thought that they would make it easier! The example is supposed to use the difference between addition and multiplication as a way to let the key affect combining the keys.

What would you suggest I use instead?

I suppose I could use something like taking the first or the last element, depending on the key.

Reply:
Maybe something like

    combine k x y = [ k ] ++ x ++ y

?

With a commutative operation you can't tell in which order the elements are being inserted which is an important gotcha here.

Reply:
I'm not a big fan of combining the key with the values in the way you suggest. That's how it's documented in Data.Map and I specifically tried to avoid it because I think it seems very contrived. 

The point about the order of operations is made in the last lines of the documentation. This mirrors the documentation of `fromListWith` which also has examples which uses commutative operators but then explains below the order of operations. I kind of like that structure.

Reply:
Hmm, the thing is that IMHO `fromListWith` and `fromListWithKey` really have this rather serious design mistake where

    fromListWith op [(k, a), (k, b)]

is not the obvious

    fromList [(k, op a b)]

but instead

    fromList [(k, op b a)]

I wish we could fix this properly but that's not a task for this PR.

So IMHO we need to ensure that potential users of `fromListWith[Key]` become aware of this gotcha ASAP so they _do not_ produce a bug, for which I feel we, as the library authors and maintainers, would share some responsibility!

Currently the only way these haddocks _hint_ at the gotcha is in the last two lines, which IMHO are fairly abstract and easy to overlook TLDR-style.

I think the documentation for `fromListWith` is less problematic because the first example with the commutative operation is very short, and the next example is very clear about the issue.

So maybe consider using David's free magma suggestion (which I feel is _slightly_ abstract) or maybe just use `Sub` and `Div` instead of `Add` and `Mul`…